### DISCUSSION TITLE
Exporting product, org.antlr.runtime version conflict problem

### DISCUSSION THREAD

**Discussion by pelva** [eclipse-sirius/sirius-desktop#351] [Q&A]
Thank you to the community for your assistance. I've managed to implement a small diagram using Sirius. However, I encountered an issue when exporting the product: I received an error stating that org.eclipse.sirius.common.acceleo.aql could not load.
```

  Unresolved requirement: Require-Bundle: org.eclipse.sirius.common.acceleo.aql
    -> Bundle-SymbolicName: org.eclipse.sirius.common.acceleo.aql; bundle-version="7.0.1.202205300817"; singleton:="true"
       org.eclipse.sirius.common.acceleo.aql [254]
         Unresolved requirement: Require-Bundle: org.eclipse.acceleo.query; bundle-version="[5.0.0,8.0.0)"
           -> Bundle-SymbolicName: org.eclipse.acceleo.query; bundle-version="7.0.0.202102190929"
              org.eclipse.acceleo.query [127]
                Unresolved requirement: Require-Bundle: org.antlr.runtime; bundle-version="[4.7.2,4.7.3)"


```
Upon inspecting the plugin list in the exported product, I found that org.antlr.runtime is using version 3.2.0, as shown below:

![image](https://github.com/eclipse-sirius/sirius-desktop/assets/135562884/b00cf303-fc2d-4898-ac2f-d51c2c30b3dc)
I am using Eclipse version 2022-06 (4.24.0), and the product configuration in Product Definition is based on plugins.
![image](https://github.com/eclipse-sirius/sirius-desktop/assets/135562884/7a4c347e-129d-46d4-91c0-d359954cc772)

In the target definition, there are indeed two versions of org.antlr.runtime.
![image](https://github.com/eclipse-sirius/sirius-desktop/assets/135562884/7058a677-32d4-4d1a-8b97-ba24d56eb74f)

But when I specified version 4.7.2, it did not work. The exported org.antlr.runtime version was still 3.2.0.
![image](https://github.com/eclipse-sirius/sirius-desktop/assets/135562884/fe00c44c-7979-40b5-8acd-9dfa9bc9dc75)

Any help would be much appreciated!

Thanks!!

**Comment by pelva** [comment]
Maybe I should configure Product as Feature Based, but it can work if I manually copy 4.7.2.

**Comment by pcdavid** [comment]
Hi.

Sirius Desktop depends on ANTLR 4.7.2 via AQL (Acceleo Query Language). Technically it's "optional", but AQL is the main/default language used for writing expressions in odesign files, so you very probably need it.

The other version, ANTLR 3.2.0 seems to be required by Xtext-related stuff (MWE2 is part of the Xtext/Xtend/Xpand/etc. stack). Neither versions of ANTLR are marked as singleton bundles, so normally it should be possible to have both versions installed at the same time inside a product (as long as there is no mix at runtime, it is normally safe).

Maybe it's a bug in PDE export which forces the use of a single version? We do not use PDE export to create our products and prefer using Tycho for that, so I'm not that familiar with this part of PDE.

Anyway, do you actually _need_ Xtext-related stuff in your modeler? If not, maybe you can tweak your product definition/export configuration to avoid these parts.

Also, do you need to export a whole product? Exporting just the plug-ins/features that make up your own modeler is usually fine. Then it can be installed in any existing product which has the required dependencies (like [Obeo Designer](https://www.obeodesigner.com/) or can get them from registered update sites (like the [Eclipse Modeling Tools Package](https://www.eclipse.org/downloads/packages/release/2024-03/r/eclipse-modeling-tools)).

**Reply by pelva** [reply]
Dear pcdavid,
I wanted to extend my heartfelt gratitude for taking the time to respond to my post. Your insight and suggestion provided an alternative perspective that I hadn't considered before. It's incredibly generous of you to share your expertise and offer a different approach to tackle the problem I was facing.

Your contribution not only helped me move forward with my project but also broadened my understanding of the issue at hand. I truly appreciate your willingness to assist fellow Sirius users like myself.

Once again, thank you for your invaluable input. Your assistance has made a significant difference, and I am truly grateful.

Best regards
